src/main/java/io/quarkus/security/identity/SecurityIdentity.java

```diff
@@ -62,16 +63,21 @@ default <T extends Principal> T getPrincipal(Class<T> clazz) {
 
     /**
      * Checks if a user has a given role. These roles must be resolvable in advance for every request.
-     * <p>
-     * If more advanced authorization support is required than can be provided by a simple role based system
-     * then {@link #checkPermission(Permission)} and {@link #checkPermissionBlocking(Permission)} should be used
-     * instead.
-     * <p>
      *
      * @return <code>true</code> if the identity has the specified role.
      */
     boolean hasRole(String role);
 
+    /**
+     * Returns the set of all permissions held by the user. These permissions must be resolvable in advance for every request.
+     * <p>
+     * This set should either be unmodifiable, or a defensive copy so attempts to change the role set do not modify
+     * the underlying identity.
+     *
+     * @return The set of all permissions held by the user
+     */
+    Set<Permission> getPermissions();
```

Comment:
This will make this change breaking for whoever implemented `SecurityIdentity`. Have you considered:

1. Define this method default?
2. Document what is relationship between `checkPermission` and `getPermission`?

Reply:
By "breaking" I mean they will have to "do something" (impl. it). I don't say it is a big deal.

Reply:
@michalvavrik As far as breaking things is concerned, I'm not too worried, we do add methods periodically here, and it all can really work only in Quarkust...

It does not conflict with `checkPermission` because `getPermissions` returns a list of possessed permissions, while `checkPermission` is about the identity finding out if one of its posssesed permissions implies the passed in required permission.

But I did present this PR the wrong way. As you made it clear to me in https://github.com/quarkusio/quarkus/issues/43717, where you also thought about `QuarkusSecurityIdentity.Builder#addPermission`, this PR does not really offers an alternative plan, all this PR is about is to let users see the possessed permissions. 
I thought Quarkus Security itself can iterate over these permissions, without dealing with permission checkers...

But we can start from `QuarkusSecurityIdentity.Builder.addPermission` only for now, thank you

Reply:
Understood. I think majority of users will just add `java.security.Permission`, in which case it makes sense that we allow to list them.